Return-Path: <root@tux>
Received: from deliver ([unix socket])
	 by tux (Cyrus 2.5.7) with LMTPA;
	 Fri, 13 May 2016 02:00:01 +0200
X-Sieve: CMU Sieve 2.4
Path: news.gmane.org!not-for-mail
From: Waiman Long <Waiman.Long@hpe.com>
Newsgroups: gmane.linux.kernel
Subject: [PATCH v3 2/4] locking/rwsem: Don't wake up one's own task
Date: Thu, 12 May 2016 18:56:49 -0400
Lines: 28
Approved: news@gmane.org
Message-ID: <1463093811-10684-3-git-send-email-Waiman.Long@hpe.com>
References: <1463093811-10684-1-git-send-email-Waiman.Long@hpe.com>
NNTP-Posting-Host: plane.gmane.org
X-Trace: ger.gmane.org 1463093868 19307 80.91.229.3 (12 May 2016 22:57:48 GMT)
X-Complaints-To: usenet@ger.gmane.org
NNTP-Posting-Date: Thu, 12 May 2016 22:57:48 +0000 (UTC)
Cc: linux-kernel@vger.kernel.org, Davidlohr Bueso <dave@stgolabs.net>,
	Jason Low <jason.low2@hp.com>,
	Dave Chinner <david@fromorbit.com>,
	Peter Hurley <peter@hurleysoftware.com>,
	Scott J Norton <scott.norton@hpe.com>,
	Douglas Hatch <doug.hatch@hpe.com>,
	Waiman Long <Waiman.Long@hpe.com>
To: Peter Zijlstra <peterz@infradead.org>,
	Ingo Molnar <mingo@redhat.com>
Original-X-From: linux-kernel-owner@vger.kernel.org Fri May 13 00:57:36 2016
Envelope-to: glk-linux-kernel-3@plane.gmane.org
Original-Received: from vger.kernel.org ([209.132.180.67])
	by plane.gmane.org with esmtp (Exim 4.69)
	(envelope-from <linux-kernel-owner@vger.kernel.org>)
	id 1b0zY0-0000sj-0o
	for glk-linux-kernel-3@plane.gmane.org; Fri, 13 May 2016 00:57:36 +0200
Original-Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
	id S1752851AbcELW5P (ORCPT <rfc822;glk-linux-kernel-3@m.gmane.org>);
	Thu, 12 May 2016 18:57:15 -0400
Original-Received: from g1t6220.austin.hp.com ([15.73.96.84]:37999 "EHLO
	g1t6220.austin.hp.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
	with ESMTP id S1751111AbcELW5N (ORCPT
	<rfc822;linux-kernel@vger.kernel.org>);
	Thu, 12 May 2016 18:57:13 -0400
Original-Received: from g2t4688.austin.hpicorp.net (g2t4688.austin.hpicorp.net [15.94.10.174])
	by g1t6220.austin.hp.com (Postfix) with ESMTP id 58CF61AF;
	Thu, 12 May 2016 22:57:11 +0000 (UTC)
Original-Received: from RHEL65.localdomain (longwa3.americas.hpqcorp.net [16.214.133.32])
	by g2t4688.austin.hpicorp.net (Postfix) with ESMTP id 781A65E;
	Thu, 12 May 2016 22:57:10 +0000 (UTC)
X-Mailer: git-send-email 1.7.1
In-Reply-To: <1463093811-10684-1-git-send-email-Waiman.Long@hpe.com>
Original-Sender: linux-kernel-owner@vger.kernel.org
Precedence: bulk
List-ID: <linux-kernel.vger.kernel.org>
X-Mailing-List: linux-kernel@vger.kernel.org
Xref: news.gmane.org gmane.linux.kernel:2221055
Archived-At: <http://permalink.gmane.org/gmane.linux.kernel/2221055>

As rwsem_down_read_failed() will queue itself and potentially call
__rwsem_do_wake(sem, RWSEM_WAKE_ANY), it is possible that a reader
will try to wake up its own task. This patch adds a check to make
sure that this won't happen.

Signed-off-by: Waiman Long <Waiman.Long@hpe.com>
---
 kernel/locking/rwsem-xadd.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/kernel/locking/rwsem-xadd.c b/kernel/locking/rwsem-xadd.c
index 7ccab5c..22f7d58 100644
--- a/kernel/locking/rwsem-xadd.c
+++ b/kernel/locking/rwsem-xadd.c
@@ -202,7 +202,8 @@ __rwsem_do_wake(struct rw_semaphore *sem, enum rwsem_wake_type wake_type)
 		 */
 		smp_mb();
 		waiter->task = NULL;
-		wake_up_process(tsk);
+		if (tsk != current)
+			wake_up_process(tsk);
 		put_task_struct(tsk);
 	} while (--loop);
 
-- 
1.7.1


